src/turbine/turbine_tree.zig

```diff
@@ -0,0 +1,804 @@
+        const gopr = try self.cache.getOrPut(epoch);
+
+        if (gopr.found_existing) {
+            if (gopr.value_ptr.alive(self.cache_entry_ttl)) {
+                return gopr.value_ptr.turbine_tree.acquireUnsafe();
+            } else {
+                gopr.value_ptr.turbine_tree.releaseUnsafe();
+                std.debug.assert(self.cache.swapRemove(epoch));
+            }
+        }
+
+        return null;
+    }
+
+    pub fn put(self: *TurbineTreeCache, epoch: Epoch, turbine_tree: *TurbineTree) !void {
+        try self.cache.put(epoch, .{
+            .created = Instant.now(),
+            .turbine_tree = turbine_tree.acquireUnsafe(),
+        });
+    }
+};
+
+/// A TurbineTree is a data structure used to determine the set of nodes to
+/// broadcast or retransmit shreds to in the network.
+pub const TurbineTree = struct {
+    allocator: std.mem.Allocator,
+    my_pubkey: Pubkey,
+    /// All staked nodes + other known tvu-peers + the node itself;
+    /// sorted by (stake, pubkey) in descending order.
+    nodes: std.ArrayList(Node),
+    /// Pubkey -> index in nodes
+    index: std.AutoArrayHashMap(Pubkey, usize),
+    /// Weighted shuffle of node stakes
+    weighted_shuffle: WeightedShuffle,
+    /// The reference count is used to facilitate deallocation, it does not
+    /// provide thread safety in a general sense.
+    reference_count: AtomicUsize,
+
+    /// The maximum number of nodes each node should retransmit to
+    pub const DATA_PLANE_FANOUT: usize = 200;
+    /// The maximum depth of the TurbineTree (0->1->2->3)
+    /// Fanout of 200 and max depth of 4 allows for ~200^3 (8 million) nodes
+    pub const MAX_TURBINE_TREE_DEPTH: usize = 4;
+    /// The maximum number of nodes per IP address
+    /// When this limit is reached, the nodes contact info is removed so that
+    /// the shuffle is deterministic but the node is not used for retransmit
+    pub const MAX_NODES_PER_IP_ADDRESS: usize = 10;
+
+    /// Nodes in the TurbineTree may be identified by solely their
+    /// pubkey if they are not in the gossip table or their contact info
+    /// is not known
+    pub const NodeId = union(enum) {
+        contact_info: ThreadSafeContactInfo,
+        pubkey: Pubkey,
+    };
+
+    /// A node in the TurbineTree
+    pub const Node = struct {
+        id: NodeId,
+        stake: u64,
+
+        pub fn pubkey(self: Node) Pubkey {
+            return switch (self.id) {
+                .contact_info => |ci| ci.pubkey,
+                .pubkey => |pk| pk,
+            };
+        }
+
+        pub fn contactInfo(self: Node) ?ThreadSafeContactInfo {
+            return switch (self.id) {
+                .contact_info => |ci| ci,
+                .pubkey => null,
+            };
+        }
+
+        pub fn tvuAddress(self: Node) ?SocketAddr {
+            return switch (self.id) {
+                .contact_info => |ci| ci.tvu_addr,
+                .pubkey => null,
+            };
+        }
```

Comment:
It would make more sense for all three of these to be methods on `NodeId`.

Reply:
I disagree, I think think given the concept of what a 'Node' is in this instance its fine to have these methods here, it would be strange to call 'node.id.tvuAddress()' in my opinion. Whenever calling these methods it is very clear that we want the nodes pubkey, contact info or tvu address. I understand your logic and we can change it if you insist but I dont think its necessary

Reply:
Here's my position: I don't feel too strongly about this either, however, depending on what you think about my proposal to use `std.MultiArrayList`, it would likely end up being more convenient and efficient. If we decide to go forward with the `std.MultiArrayList` idea, it would mean you only need to fetch one field in order to call these methods, instead of having to construct the entire `Node` to do so.

So I would like a resolution on that and related conversations before concluding this; if you think the `std.MultiArrayList` proposal isn't feasible, or is out of scope, that's acceptable, and I'd fine with whatever decision you make here.

On the side: I think that scoping data tightly is beneficial for understanding how things interact; there's no reason for `tvuAddress` to have access to the other fields in the `Node`, so why would you need all of it in order to fetch that specific data? Ergo in general I lean towards scoping methods rather than "conceptual units", but I won't die on this hill either.

Reply:
7e9c4499